### PR TITLE
Tour of Beam terraform scripts for back-end deployment

### PR DESCRIPTION
**Pull request by ruslan-ikhsan**
PR per https://github.com/apache/beam/issues/23475

**Description:** 
Need to create terraform scripts for Tour of Beam backend deployment to GCP project:
• scripts for deployment ToB cloud functions to GCP
• scripts for service accounts if needed

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: learning/tour-of-beam/terraform/variables.tf [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#Buckets
+
+variable "function_bucket_name" {
+  description = "Name of Bucket to Store Cloud Functions"
+  default     = "tour-of-beam-cloudfunction-bucket"
+}
+
+variable "function_bucket_location" {
+  description = "Cloud Functions Bucket Region"
+  default     = "US"
```
Comment (olehborysevych) [inline learning/tour-of-beam/terraform/variables.tf:27]: is this a correct default value for location?

File: learning/tour-of-beam/terraform-v2/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    - Service Account User
+    - Storage Admin
+    - Storage Object Admin
+- Exported JSON Key of service account above
+
+#### Configuring your environment
+
+Steps below will:
+1. Enable required APIs for the project
+2. Create service account and assign required IAM roles to it (service account to run the function with)
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/README.md:22]: Usually we deploy infrastructure under some Service Account already, maybe we don't create another?

File: learning/tour-of-beam/terraform-v2/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+```bash
+# Create environment directory per your requirements/policy
+mkdir environments/dev
+cd ../environments/dev
+# Import exported JSON Key using
+export GOOGLE_APPLICATION_CREDENTIALS = '..key path...'
+# Initiliaze and run terraform
+terraform init
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/README.md:34]: How do we integrate it with Playground terraform deployment
CF depends on Datastore which is deployed in playground/terraform/infrastructure

File: learning/tour-of-beam/terraform-v2/environments/uat/main.tf [hunk 4]
```diff
@@ -0,0 +1,37 @@
+locals {
+  env = "uat"
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/environments/uat/main.tf:2]: NIT  what is `uat`?

File: learning/tour-of-beam/terraform-v2/environments/uat/main.tf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  bucket_name = var.bucket_name
+  depends_on = [module.iam, module.api_enable]
+}
+
+module "api_enable" {
+  source = "../../modules/api_enable"
+  project_id = var.project_id
+}
+
+module "cloud_functions" {
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/environments/uat/main.tf:28]: why do we keep it in `environments`?
I thought `environments` is where only runtime variables defined

File: learning/tour-of-beam/terraform-v2/README.md (hunk 3 above)
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/README.md:34]: NIT Is it initialized with GCS terraform backend here?

File: learning/tour-of-beam/terraform-v2/environments/uat/variables.tf [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  description = "Bucket name to store function source code"
+}
+variable "project_id" {
+  description = "Our GCP Project"
+}
+variable "service_account_id" {
+  description = "Name of SA to run Cloud Function"
+}
+variable "region" {
+  default = "europe-west1"
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/environments/uat/variables.tf:11]: Bad default choice, it is better be the same as in Playground: `us-central1`

File: learning/tour-of-beam/terraform-v2/modules/cloud_functions/main.tf [hunk 7]
```diff
@@ -0,0 +1,34 @@
+resource "google_cloudfunctions_function" "cloud_function" {
+  count                 = length(var.entry_point_names)
+  name                  = "${var.entry_point_names[count.index]}"
+  runtime               = "go116"
+  project               = var.project_id
+  service_account_email = var.service_account_id
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/modules/cloud_functions/main.tf:6]: Is this the only reason we created a SA?

Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/modules/cloud_functions/main.tf:6]: Sorry, what reason?
In this context, we need this SA cloud functions will run with it

File: learning/tour-of-beam/terraform-v2/environments/uat/variables.tf (hunk 6 above)
Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/environments/uat/variables.tf:11]: Action: To rename default "region"

File: learning/tour-of-beam/terraform-v2/environments/uat/main.tf (hunk 5 above)
Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/environments/uat/main.tf:28]: That's correct, my mistake. 
Action: To move files out of "environments"

File: learning/tour-of-beam/terraform-v2/environments/uat/main.tf (hunk 4 above)
Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/environments/uat/main.tf:2]: That is redundant
Action: shall be removed

File: learning/tour-of-beam/terraform-v2/README.md (hunk 3 above)
Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/README.md:34]: Yes, initialized with pre-created GCS bucket

File: .github/workflows/gcp-deploy.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    - name: 'Set up Cloud SDK'
+      uses: 'google-github-actions/setup-gcloud@v0'
+
+    - name: 'Use gcloud CLI'
+      run: 'gcloud info'
+
+    - name: 'Start Google Cloud Build trigger'
+      run: |
+        export CLOUDSDK_CORE_DISABLE_PROMPTS=1
+        gcloud beta builds triggers run ${{ secrets.GCP_BUILD_STAGING_TRIGGER_ID }} --tag=${{ github.event.release.tag_name }}
```
Comment (eantyshev) [inline .github/workflows/gcp-deploy.yml:26]: Is it synchronous? does it wait for result?

File: learning/tour-of-beam/terraform-v2/backend.tf [hunk 9]
```diff
@@ -0,0 +1,6 @@
+terraform {
+  backend "gcs" {
+    bucket = "tour-of-beam-backend-tfstate-bucket"
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/backend.tf:3]: Could we load this from a variable?

File: learning/tour-of-beam/terraform-v2/backend.tf [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+    prefix = "env/uat"
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/backend.tf:4]: is it relevant? I mean `uat`

File: learning/tour-of-beam/terraform-v2/cloudbuild.yaml [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+steps:
+  - id: 'branch name'
+    name: 'alpine'
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/cloudbuild.yaml:17]: does it spin up a docker instance just to run a branch name?

File: learning/tour-of-beam/terraform-v2/cloudbuild.yaml [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    name: 'hashicorp/terraform:1.3.1'
+    entrypoint: 'sh'
+    args:
+      - '-c'
+      - |
+        if [ -d "environments/$BRANCH_NAME/" ]; then
+          cd environments/$BRANCH_NAME
+          terraform init
+        else
+          for dir in environments/*/
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/cloudbuild.yaml:36]: what's the usage scenario here? (no $BRANCH_NAME env dir)

File: learning/tour-of-beam/terraform-v2/modules/cloud_functions/main.tf [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  # Get the source code of the cloud function as a Zip compression
+  trigger_http = true
+  # Name of the function that will be executed when the Google Cloud Function is triggered
+  entry_point = var.entry_point_names[count.index]
+
+  environment_variables = {
+    DATASTORE_PROJECT_ID="test-cloud-func-deploy"
+    TOB_MOCK=1
+  }
+
```
Comment (eantyshev) [inline learning/tour-of-beam/terraform-v2/modules/cloud_functions/main.tf:20]: let's also make a variable for `available_memory_mb`

File: learning/tour-of-beam/terraform-v2/README.md (hunk 2 above)
Comment (olehborysevych) [reply learning/tour-of-beam/terraform-v2/README.md:22]: I think that Ruslan is correct. We need to have a separate SA for cloud functions to follow the minimum priviledge principle

File: learning/tour-of-beam/terraform-v2/modules/iam/main.tf [hunk 14]
```diff
@@ -0,0 +1,15 @@
+resource "google_service_account" "cloud_function_sa" {
+  account_id   = var.service_account_id
+  display_name = "Service Account to run Cloud Functions"
+}
+
+resource "google_project_iam_member" "terraform_service_account_roles" {
+  for_each = toset([
+    "roles/cloudfunctions.admin", "roles/storage.objectViewer",
+    "roles/storage.objectCreator", "roles/iam.serviceAccountUser"
```
Comment (olehborysevych) [inline learning/tour-of-beam/terraform-v2/modules/iam/main.tf:9]: Not sure why cloud functions need storage roles

File: playground/terraform/environment/beta/state.tfbackend [hunk 15]
```diff
@@ -17,4 +17,4 @@
 # under the License.
 #
 
-bucket  = "beam_playground_terraform"
```
[the remark this answers is not in the document]
Comment (olehborysevych) [reply playground/terraform/environment/beta/state.tfbackend:20]: Do we really intend to delete this file?

**Review by olehborysevych — CHANGES_REQUESTED** [review]

File: learning/tour-of-beam/terraform-v2/README.md [hunk 16]
```diff
@@ -0,0 +1,60 @@
+# [Tour of Beam] Deploying backend infrastructure as GCP Cloud Function with GCP, Terraform and Cloud Build
+This tutorial explains how to create Tour of Beam backend infrastructure as a GCP Cloud Functions using Infrastructure as a Code (IaC) and GitOps methodoliges.
```
[the remark this answers is not in the document]
Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/README.md]: Readme changed

File: learning/tour-of-beam/terraform-v2/README.md [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+3. Create bucket to archive and store source code
+4. Create cloud functions to each function defined in backend source code
+
+
+```bash
+# Create environment directory per your requirements/policy
+mkdir environments/dev
+cd ../environments/dev
+# Import exported JSON Key using
+export GOOGLE_APPLICATION_CREDENTIALS = '..key path...'
```
[the remark this answers is not in the document]
Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/README.md]: This is fixed by using gcloud init

File: learning/tour-of-beam/terraform-v2/modules/iam/main.tf (hunk 14 above)
Comment (ruslan-ikhsan) [reply learning/tour-of-beam/terraform-v2/modules/iam/main.tf]: Correct. This is redundant, removed

**Pull request closed without merging**